### PR TITLE
validation: add hostname validation

### PR DESCRIPTION
**Pull request by Mashimiao**
Signed-off-by: Ma Shimiao <mashimiao.fnst@cn.fujitsu.com>

### CODE REVIEW

File: validate.go [hunk 1]
```diff
@@ -163,9 +165,15 @@ func checkLinux(spec rspec.Linux, rootfs string) {
 	for index := 0; index < len(spec.Namespaces); index++ {
 		if !namespaceValid(spec.Namespaces[index]) {
 			logrus.Fatalf("Namespace %s is invalid.", spec.Namespaces[index])
+		} else if spec.Namespces[index].Type == spec.UTSNamespace {
+			utsExist = true
```
Comment (liangchenye) [inline validate.go]: s/spec.UTS/rspec.UTS

File: validate.go [hunk 2]
```diff
@@ -163,6 +163,8 @@ func checkLinux(spec rspec.Linux, rootfs string) {
 	for index := 0; index < len(spec.Namespaces); index++ {
 		if !namespaceValid(spec.Namespaces[index]) {
 			logrus.Fatalf("Namespace %s is invalid.", spec.Namespaces[index])
+		} else if spec.Namespaces[index].Type == rspec.UTSNamespace && rspec.Spec.Hostname != "" {
```
Comment (wking) [inline validate.go]: This probably needs the whole rspec.Spec, not its current rspec.Linux.  Then you can check `spec.Hostname` and `spec.Linux.Namespaces`.  With cross-platform options like ‘hostname’ depending on platform-specific options like UTS namespaces, this kind of cross-cutting is unavoidable.

**Comment by liangchenye** [comment]
LGTM

**Comment by liangchenye** [comment]
Hi @Mashimiao , the [spec](https://github.com/opencontainers/runtime-spec/blob/master/config.md#hostname) says:
```
 On Linux, you can only set this if your bundle creates a new UTS namespace.
```
So host name could be empty even a config file has a UTS namespace.
We should get an error when  hostname is empty and no UTS namespace is created

File: validate.go [hunk 3]
```diff
@@ -163,6 +163,8 @@ func checkLinux(spec rspec.Linux, rootfs string) {
 	for index := 0; index < len(spec.Namespaces); index++ {
 		if !namespaceValid(spec.Namespaces[index]) {
 			logrus.Fatalf("Namespace %s is invalid.", spec.Namespaces[index])
+		} else if spec.Namespaces[index].Type == rspec.UTSNamespace && hostname != "" {
```
Comment (liangchenye) [inline validate.go]:
```
 On Linux, you can only set this if your bundle creates a new UTS namespace.
```
So host name could be empty even a config file has a UTS namespace.
We should get an error when hostname is empty and no UTS namespace is created

**Comment by liangchenye** [comment]
add a WIP link in https://github.com/opencontainers/ocitools/issues/66

File: validate.go (hunk 3 above)
Comment (wking) [inline validate.go]: On Thu, May 19, 2016 at 08:30:00PM -0700, 梁辰晔 (Liang Chenye) wrote:
> So host name could be empty even a config file has a UTS namespace.

Correct.  Maybe the UTS namespace is for a NIS domain name.

> We should get an error when hostname is empty and no UTS namespace
> is created

I think you mean, when the hostname is *not* empty and no UTS
namespace is created.  All other combinations (UTS and no hostname,
neither UTS not hostname, both UTS and hostname) are valid.

Comment (liangchenye) [inline validate.go]: Yes, `when hostname is not empty`.
@Mashimiao 's first commit did the right thing and I mislead him to squash code lines.

**Comment by Mashimiao** [comment]
@liangchenye  @wking   fixed.

**Comment by wking** [comment]
b62e1cc5 looks good to me.

**Comment by liangchenye** [comment]
LGTM

File: validate.go [hunk 4]
```diff
@@ -152,7 +152,9 @@ func checkProcess(process rspec.Process, rootfs string) {
 }
 
 //Linux only
-func checkLinux(spec rspec.Linux, rootfs string) {
+func checkLinux(spec rspec.Linux, hostname string, rootfs string) {
+	utsExist := false
```
Comment (mrunalp) [inline validate.go]: nit: `utsExist` --> `utsExists`

Comment (Mashimiao) [inline validate.go]: fixed

File: validate.go [hunk 5]
```diff
@@ -177,9 +179,15 @@ func checkLinux(spec rspec.Linux, rootfs string) {
 	for index := 0; index < len(spec.Namespaces); index++ {
 		if !namespaceValid(spec.Namespaces[index]) {
 			logrus.Fatalf("namespace %v is invalid.", spec.Namespaces[index])
+		} else if spec.Namespaces[index].Type == rspec.UTSNamespace {
+			utsExists = true
 		}
 	}
 
+	if !utsExists && hostname != "" {
+		logrus.Fatalf("New UTS namespace should to be created.")
```
Comment (mrunalp) [inline validate.go]: "hostname requires a new UTS namespace to be specified as well"

Comment (Mashimiao) [inline validate.go]: @mrunalp  fixed.

**Comment by mrunalp** [comment]
LGTM

**Pull request merged**